### PR TITLE
Generate a unique jobId for related Extraction messages

### LINKED ISSUE #28: Extraction message start Date should not be an Option
**Is your feature request related to a problem? Please describe.**
The `start` field should not be optional, since we always know the time that a message came through. Furthermore, having `start` as an `Option[Date]` makes it difficult to sort the resulting extraction lists by Date, as the built-in helpers (e.g. `.sortWith(_.after(_))`) no longer apply.

**Describe the solution you'd like**
`start` should be a `Date` instead of an `Option[Date]`

**Describe alternatives you've considered**
`.sortWith(_.getOrElse(epochDate).after(_.getOrElse(epochDate)))` might work, but is a bit lengthy and potentially error-prone

**Additional context**
The context here was for updating the File/Dataset "Extractions" tab to display a Job ID. Some sorting was necessary to reorder things by Date. We might also consider getting rid of the `end` Date, since it is widely unused.

See https://github.com/clowder-framework/clowder/blob/develop/app/models/Extraction.scala

### PR DESCRIPTION
**Pull request by bodom0015**
## Description
Relates to https://github.com/clowder-framework/pyclowder/pull/7
Fixes #28 

Demonstration: http://recordit.co/BThq8PnxM8

### Problem
As an extractor developer or user, it would be nice to be able to track which jobs generated which status messages. It is difficult to track and debug using ExtractorMessages because for a given file + extractor, if multiple "jobs" or "runs" of the extractor are present, it is hard to determine which messages came from which jobs.

### Approach
* Generate a unique/random UUID when submitting a file to an extractor. We can then forward this jobId to pyClowder, who can send it back in subsequent status updates. Ultimately, this is saved in the database in a new field.
* Rework the Extractions tab in the UI to categorize by jobId
* Sort newest to oldest for top 2 levels (extractors and jobs), but sort oldest to newest for actual event list for each job

### How to Test
1. Clone my pyclowder branch locally and run ./docker.sh to build the images
    * `git clone https://github.com/clowder-framework/pyclowder pyclowder-testing
    * `cd pyclowder-testing/`
    * `./docker.sh`
2. Run the wordcount extractor from the custom pyclowder image you just built
    * `docker run -it --rm --net=host -e REGISTRATION_ENDPOINTS="${REGISTRATION_ENDPOINTS}" -e RABBITMQ_URI="${RABBITMQ_URI}" clowder/extractors-wordcount:latest`
    * NOTE: You should not see Docker pull an image when running this command
3. Start up RabbitMQ + enable RabbitMQ plugin in Clowder config
4. Checkout and run this branch locally
5. Login to Clowder
6. Create a dataset, if you haven't already, and upload a text file to it
    * You can use `echo 'hello world' > hello.json` to create a simple text file
    * You should see the extractor begin to process the file
    * You should see in the extractor logs that check_message outputted a `jobid`
7. Open a mongo shell and query the contents of the `extractions` by job_id:
    * `docker exec -it mongo bash`
    * `mongo`
    * `use clowder`
    * `db.extractions.find({ 'job_id': ObjectId('YOUR_JOBID_HERE') })
    * You should see that only extractions associated with this particular "run" are shown (on this particular extractor + file)
8. View the file in Clowder, check the Extractions tab, and expand an extractor
    * You should see a list of all extractors / events that have run on the file are listed here
    * Upon expanding, you should see a list of clickable unique jobs shown here - note that the last update time is displayed instead of the Job ID
    * You should see a chevron icon point to the right of each job indicating that an expandable row is collapsed
9. Click a job to expand it
    * You should see the chevron icon point downward to indicate the row is expanded
    * You should see all events from that job

## Review Time Estimate

- [ ] Immediately
- [x] Within one week
- [ ] When possible

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue) 
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the CHANGELOG.md.
- [ ] I have signed the CLA
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by max-zilla** [comment]
I ran your test and although I encountered unrelated network issue in actually counting the words, the job ID was there in pyclowder & mongo and displayed on GUI as expected.

File: app/services/mongodb/MongoDBExtractionService.scala [hunk 1]
```diff
@@ -118,7 +132,7 @@ class MongoDBExtractionService extends ExtractionService {
           }
           case None => {}
         }
-        groupings = groupings + (e.extractor_id -> ExtractionGroup(grp_start.toString, grp_end.toString, grp_endmsg, grp_list :+ e))
+        groupings = groupings + (e.extractor_id -> ExtractionGroup(grp_start.toString, grp_end.toString, grp_endmsg, grp_map))
```
Comment (bodom0015) [inline app/services/mongodb/MongoDBExtractionService.scala:135]: I think I've made a mistake in the logic for building up the ExtractionGroup Map here... the UI appears to only show one event from each job, and no longer lists all events related to the given jobId.

that last bit of syntax with the `:+` is probably some Scala shorthand to concatenate the new entry into the Map

**Comment by bodom0015** [comment]
So I am confused... I've pushed 2 new commits to this branch and they don't appear to show on this PR (yet?)... but if I click over to the [source branch](https://github.com/clowder-framework/clowder/tree/extractor-job-tracking-improvements) in GitHub, I can see the 2 new commits are there.

I am not sure why the new commits aren't showing up on this PR as expected, but may need to close this and reopen it if they refuse to appear.

**Comment by bodom0015** [comment]
There we go. @max-zilla's suggestion to change the base branch worked, and now the PR shows the missing commits.

I've updated this PR to fix the single job-event bug, as well as to include a better display for the Job ID in the UI. Instead of just a new column showing the ID, each unique Job ID adds an expandable row in the subtable. This way, the user can choose an extractor to see all jobs run by that extractor, and select a particular job to see extraction events or status updates from that job.

There are still some open issues regarding how to sort the information in the table, and additional PRs will likely be required to address these concerns

**Comment by lmarini** [comment]
General question, what happens if we deploy this new features and we have extractors that don't support it?

File: app/views/extractions/extractionsList.scala.html [hunk 2]
```diff
@@ -76,3 +102,28 @@
             });
         }
 </script>
+<style>
```
[the remark this answers is not in the document]
Comment (bodom0015) [reply app/views/extractions/extractionsList.scala.html:103]: Absolutely.. Would you have any recommendation of where to move them? I see there is a `public/stylesheets/` directory. Would this be an appropriate place to house these styles?

File: app/services/mongodb/MongoDBExtractionService.scala [hunk 3]
```diff
@@ -77,37 +81,46 @@ class MongoDBExtractionService extends ExtractionService {
...
         var grp_end = format.parse(groupings(e.extractor_id).latestMsgTime)
         var grp_endmsg = groupings(e.extractor_id).latestMsg
-        var grp_list = groupings(e.extractor_id).allMsgs
+
+        // For each job_id, keep track of all extraction events for that job
+        var grp_map = groupings(e.extractor_id).allMsgs
+        if(grp_map.contains(e.job_id)) {
+          val existing: List[Extraction] = grp_map(e.job_id)
+          val updated = List(e) ++ existing
+          // Re-sort list by timestamp (newest first)
```
[the remark this answers is not in the document]
Comment (bodom0015) [reply app/services/mongodb/MongoDBExtractionService.scala:102]: Good catch! Not anymore, will remove it

File: app/api/Extractions.scala [hunk 4]
```diff
@@ -129,29 +130,31 @@ class Extractions @Inject()(
...
         if (UUID.isValid(id.stringify)) {
           files.get(id) match {
             case Some(file) => {
-              current.plugin[RabbitmqPlugin].foreach {
-                // FIXME dataset not available?
-                _.fileCreated(file, None, Utils.baseUrl(request), request.apiKey)
+              // FIXME dataset not available?
+              plugin.fileCreated(file, None, Utils.baseUrl(request), request.apiKey) match {
+                case Some(jobId) => {
+                  Ok(Json.obj("status" -> "Ok", "job_id" -> jobId))
```
[the remark this answers is not in the document]
Comment (bodom0015) [reply app/api/Extractions.scala:145]: Will go ahead and fix the capitalization there. Consistency is important :)

File: app/views/extractions/extractionsList.scala.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (bodom0015) [reply app/views/extractions/extractionsList.scala.html]: I went ahead and moved these styles to `public/stylesheets/extractions.css`

**Comment by bodom0015** [comment]
I've pushed a fix for the backward compatibility issues, as well as another commit with some more explicit text for this case as discussed at this morning's Clowder meeting.

Below is a screenshot of this case and the new text. Please let me know if this might require any further changes
Screen Shot 2020-07-31 at 1 10 21 PM

**Pull request merged**